atomic_reactor/plugins/pre_add_image_content_manifest.py

```diff
@@ -0,0 +1,208 @@
+        }
+      ]
+    }
+
+    WITH content_sets specified:
+
+    {
+      "metadata": {
+        "icm_version": 1,
+        "icm_spec": "https://link.to.icm.specification",
+        "image_layer_index": 2
+      },
+      "content_sets": [
+          "rhel-8-for-x86_64-baseos-rpms",
+          "rhel-8-for-x86_64-appstream-rpms"
+      ],
+      "image_contents": [
+        {
+          "purl": "pkg:golang/github.com%2Frelease-engineering%2Fretrodep%2Fv2@v2.0.2",
+          "dependencies": [{"purl": "pkg:golang/github.com%2Fop%2Fgo-logging@v0.0.0"}],
+          "sources": [{"purl": "pkg:golang/github.com%2FMasterminds%2Fsemver@v1.4.2"}]
+        }
+      ]
+    }
+    """
+    key = PLUGIN_ADD_IMAGE_CONTENT_MANIFEST
+    is_allowed_to_fail = False
+    minimal_icm = {
+        'metadata': {
+            'icm_version': 1,
+            'icm_spec': ('https://raw.githubusercontent.com/containerbuildsystem/atomic-reactor/'
+                         'master/atomic_reactor/schemas/content_manifest.json'),
+            'image_layer_index': -1,
+        },
+        'content_sets': [],
+        'image_contents': [],
+    }
+
+    def __init__(self, tasker, workflow, remote_source_icm_url=None, destdir=IMAGE_BUILD_INFO_DIR):
+        """
+        :param tasker: ContainerTasker instance
+        :param workflow: DockerBuildWorkflow instance
+        :param icm_url: str, URL of the ICM from the Cachito request.
+        :param destdir: image path to carry content_manifests data dir
+        """
+        super(AddImageContentManifestPlugin, self).__init__(tasker, workflow)
+        self.content_manifests_dir = os.path.join(destdir, 'content_manifests', '')
+        self.icm_url = remote_source_icm_url
+        self.dfp = df_parser(self.workflow.builder.df_path, workflow=self.workflow)
+        labels = Labels(self.dfp.labels)
+        _, image_name = labels.get_name_and_value(Labels.LABEL_TYPE_COMPONENT)
+        _, image_version = labels.get_name_and_value(Labels.LABEL_TYPE_VERSION)
+        _, image_release = labels.get_name_and_value(Labels.LABEL_TYPE_RELEASE)
+        self.icm_file_name = '{}-{}-{}.json'.format(image_name, image_version, image_release)
+        self.content_sets = []
+        self._cachito_verify = True
+        self._layer_index = None
+        self._icm = None
+
+    @property
+    def layer_index(self):
+        if not self._layer_index:
+            # Default layer index is 1, because base and 'FROM scratch' images
+            #     *always* have 2 layers
+            self._layer_index = 1
+        if not base_image_is_scratch(self.dfp.baseimage):
+            inspect = self.workflow.builder.base_image_inspect
+            self._layer_index = len(inspect[INSPECT_ROOTFS][INSPECT_ROOTFS_LAYERS])
+        return self._layer_index
+
+    @property
+    def cachito_verify(self):
+        if self._cachito_verify is True:
+            cachito_conf = get_cachito(self.workflow) or {}
```

Comment:
If I'm not mistaken, `get_cachito()` raises `KeyError` when cachito is not configured. You may have to use a try/except, or use `get_cachito(self.workflow, fallback={})` (but get_cachito does not currently accept the `fallback` param, you would have to update it)

Reply:
Good catch!